.grit/patterns/sql_alchemy_v2.md

```diff
@@ -86,13 +86,105 @@ pattern no_nested_with_expr() {
     }
 }
 
+pattern get_directly_on_session() {
+    `$sess.query($cls).get($id)` => `$sess.get($cls, $id)`
+}
+
+pattern cmp_statements() {
+    r"^.+ (?:(?:==)|(?:>=)|(?:<=)|>|<) .+(?: (?:and)|(?:or) .+ (?:(?:==)|(?:>=)|(?:<=)|>|<) .+)*$",
+}
+
+pattern migrate_old_select_args($call_chain, $args, $new_stmt) {
+    $_ where {
+        if ($args <: contains list(elements=$items)) {
+            $call_chain += `select($items)`,
+        } else {
+            $call_chain += `select()`,
+        },
+        $args <: maybe contains keyword_argument(name=`select_from`, value=$table) where {
+            $call_chain += `select_from($table)`,
+        },
+        $args <: maybe contains keyword_argument(name=`order_by`, value=$col) where {
+            $call_chain += `order_by($col)`,
+        },
+        $args <: maybe contains cmp_statements() as $cmp_stmt where {
+            $call_chain += `where($cmp_stmt)`,
+        },
+        $new_stmt = join(list = $call_chain, separator = "."),
+    }
+}
+
+pattern migrate_pre_1_4_select() {
+    $call_chain = [],
+    or {
+        `select($args)`,
+        `$table.select($args)` where {
+            $call_chain += `$table`,
+        },
+    } as $stmt where or {
+        $args <: contains list(elements=$_),
+        $args <: contains keyword_argument(name=$_, value=$_),
+        $args <: contains cmp_statements(),
+    },
+    migrate_old_select_args($call_chain, $args, $new_stmt),
+    $stmt => `$new_stmt`,
```

Comment:
Idiomatically, we would make migrate_old_select args a pattern that applies to the $stmt and handles the rewrite.

```suggestion
        $stmt <: migrate_old_select_args($call_chain, $args),
```